gammapy/spectrum/energy_group.py

```diff
@@ -0,0 +1,534 @@
+
+    def apply_energy_binning(self, ebounds):
+        """TODO: document"""
+        for energy_range in EnergyRange.list_from_ebounds(ebounds):
+            list_idx_min, list_idx_max = self.find_list_idx_range(energy_range)
+
+            # Be sure to leave underflow and overflow bins alone
+            # TODO: this is pretty ugly ... make it better somehow!
+            list_idx_min = self.clip_to_valid_range(list_idx_min)
+            list_idx_max = self.clip_to_valid_range(list_idx_max)
+
+            self.make_and_replace_merged_group(
+                list_idx_min=list_idx_min,
+                list_idx_max=list_idx_max,
+                bin_type='normal',
+            )
+
+    def clip_to_valid_range(self, list_idx):
+        """TODO: document"""
+        if self[list_idx].bin_type == 'underflow':
+            list_idx += 1
+
+        if self[list_idx].bin_type == 'overflow':
+            list_idx -= 1
+
+        if (list_idx < 0) or (list_idx >= len(self)):
+            raise IndexError('Oh no!')
+
+        return list_idx
+
+    def make_and_replace_merged_group(self, list_idx_min, list_idx_max, bin_type):
+        """TODO: document"""
+        # Create a merged group object
+        group = self.make_merged_group(
+            list_idx_min=list_idx_min,
+            list_idx_max=list_idx_max,
+            bin_type=bin_type,
+        )
+
+        # Delete previous groups
+        [self.pop(list_idx_min) for _ in range(list_idx_max - list_idx_min + 1)]
+
+        # Insert the merged group
+        self.insert(list_idx_min, group)
+        self.reindex_groups()
+
+    def reindex_groups(self):
+        """TODO: document"""
+        for energy_group_idx, group in enumerate(self):
+            group.energy_group_idx = energy_group_idx
+
+    def make_merged_group(self, list_idx_min, list_idx_max, bin_type):
+        """TODO: document"""
+        left_group = self[list_idx_min]
+        right_group = self[list_idx_max]
+
+        return SpectrumEnergyGroup(
+            energy_group_idx=INVALID_GROUP_INDEX,
+            bin_idx_min=left_group.bin_idx_min,
+            bin_idx_max=right_group.bin_idx_max,
+            bin_type=bin_type,
+            energy_min=left_group.energy_min,
+            energy_max=right_group.energy_max,
+        )
+
+    def __str__(self):
+        ss = 'SpectrumEnergyGroups:\n'
+        ss += '\nInfo including underflow- and overflow bins:\n'
+        ss += 'Number of groups: {}\n'.format(len(self))
+        ss += 'Bin range: {}\n'.format(self.bin_idx_range)
+        ss += 'Energy range: {}\n'.format(self.energy_range)
+
+        # TODO
+        # ss += '\nInfo excluding underflow- and overflow bins:\n'
+        # ss += 'Number of groups: {}\n'.format(len(self))
+        # ss += 'Bin range: {}\n'.format(self.bin_idx_range)
+        # ss += 'Energy range: {}\n'.format(self.energy_range)
+
+        return ss
+
+
+class EnergyRange(object):
```

Comment:
I don't think this is the appropriate place for this class. It reproduces a lot of the functionality that is already in the [EnergyBounds](http://docs.gammapy.org/en/latest/api/gammapy.utils.energy.EnergyBounds.html#gammapy.utils.energy.EnergyBounds). See also
http://docs.gammapy.org/en/latest/utils/index.html#energy-handling-in-gammapy